src/ast.rs

```diff
@@ -446,6 +446,16 @@
     }
 }
 
+/// This is for optional values and for lists.
+trait CheckParse {
+    /// A true return means we should descend into the production, false means stop.
+    fn can_be_parsed(pop: &mut Populator<'_>) -> bool;
+}
+
+trait StaticTokenInfo {
+    const ALLOWED_TOKENS: &'static [ParseTokenType];
```

Comment:
FWIW the reason why types like `enum Type` are at the bottom is purely historical.
They should probably be closer to the top of the file.

Reply:
Sorry about the fine-grained commits.
The initial position of (some of?) my new items was mostly a result of being overwhelmed by the contents of this module and wanting to just get to prototyping.